proposals/NNNN-codable-error-printing.md

```diff
@@ -0,0 +1,175 @@
+Path: [0]/home/country
+```
+
+(Note: the output could be further improved by modifying `JSONDecoder` to write a better debug description. See [Future Directions](#future-directions) for more.)
+
+### Structure
+
+1. Description using information we know from the associated values of the error enum itself.
+1. The debug description that was passed to the error, if it is not empty.
+1. The underlying error, if it is non-nil.
+1. The coding path, neatly formatted, if it is non-empty. String keys are presented as-is, and numeric indices are presented in square brackets like `[2]` to differentiate them from string keys.
+
+More complete examples of the before/after diffs are available in the description of the pull request: https://github.com/swiftlang/swift/pull/80941.
+
+The path formatting is especially improved. Comparing the examples from above:
+
+```diff
+-[_CodingKey(stringValue: "Index 0", intValue: 0), CodingKeys(stringValue: "home", intValue: nil), CodingKeys(stringValue: "country", intValue: nil)]
++Path: [0]/home/country
+```
+
+## Detailed design
+
+```swift
+@available(SwiftStdlib 6.2, *)
+extension EncodingError: CustomDebugStringConvertible {
+  public var debugDescription: String {...}
+}
+
+@available(SwiftStdlib 6.2, *)
+extension DecodingError: CustomDebugStringConvertible {
+  public var debugDescription: String {...}
+}
+```
+
+## Source compatibility
+
+The new conformance changes the result of converting an `EncodingError` or `DecodingError` value to a string. This changes observable behavior: code that attempts to parse the result of `String(describing:)` or `String(reflecting:)` can be misled by the change of format.
+
+However, the documentation of these interfaces explicitly state that when the input type conforms to none of the standard string conversion protocols, then the result of these operations is unspecified.
+
+Changing the value of an unspecified result is not considered to be a source incompatible change.
+
+## ABI compatibility
+
+The proposal retroactively conforms two previously existing standard types to a previously existing standard protocol. This is technically an ABI breaking change: on ABI-stable platforms, we may have preexisting Swift binaries that assume that `EncodingError is CustomDebugStringConvertible` or `DecodingError is CustomDebugStrinConvertible` returns `false`, or ones that are implementing this conformance on their own.
+
+We do not expect this to be an issue in practice.
+
+## Implications on adoption
+
+[Unsure what to add here. I see stuff in [SE-0445](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0445-string-index-printing.md#implications-on-adoption), but I'm not sure how much of that applies here. I don't know if I need to be doing the `@backDeployed` things that proposal mentions, and when I look at the code from the PR, I see `@_alwaysEmitIntoClient // FIXME: Use @backDeployed`.]
```

Comment:
I suspect that in SE-0445 the back-deploying was important because the non-API public symbol `_description` existed on `String.Index` in prior Swift versions, and was reimplemented using `debugDescription`? I'm not certain that this is quite as relevant here, which suggests that there's nothing really useful to add to this section.

@lorentey can you provide any background on SE-0445 and/or advice here?